traffic_ops/experimental/webfront/webfront.go

```diff
@@ -122,188 +145,311 @@ func main() {
+}
+
+// parseRoutes reads route definitions from file, constructs the route auth handler,
+// and returns the resultant routes.
+func parseRoutes(file string) ([]*Route, error) {
+
+	// If the rule defines a routes file, we load the routes and enforce access.
+	// Routes than are not present in this file are forbidden.
 
-		Logger.Printf("%v %v Valid token. Subject=%v, ExpiresAt=%v, Capabilities=%v, Required=%v, Authorized=%v", 
-			r.Method, r.URL.RequestURI(), claims.Subject, claims.ExpiresAt, claims.Capabilities, 
-			rule.Capabilities[r.Method], isAuthorized)
+	// Note that there is currently no mechanism to trigger an update on a change in the route files. 
+	// To trigger an update, one needs to touch rules.json 
 
-	} else {
-		isAuthorized = true
+	cf, err := os.Open(file)
+	if err != nil {
+		return nil, err
+	}
+	defer cf.Close()
+
+	Logger.Printf("Loading routes file: %s", file)
+
+	var routes []*Route
+	if err := json.NewDecoder(cf).Decode(&routes); err != nil {
+		return nil, err
 	}
 
-	if isAuthorized {
-		if h := rule.handler; h != nil {
-			h.ServeHTTP(w, r)
+	for _, r := range routes {
+
+		/*
+		// If the match ends with a slash, it is treated as a prefix. 
+		// If not, it is an exact match
+		if !strings.EndsWith(r.Match, "/") {
+			r.Match = r.Match + '$'
+		}
+		*/
+
+		r.matchRegexp, err = regexp.Compile(r.Match + "$")
+		if err != nil {
+			Logger.Printf("Skip route %s ERROR: %s", r.Match, err)
+			continue
+		}
+
+		// Logger.Printf("Loaded route: %s", r.Match)
+	}
+
+	return routes, nil
+}
+
+// makeHandler constructs the appropriate Handler for the given FwdRule.
+func makeHandler(r *FwdRule) (http.Handler, error) {
+
+	if h := r.Forward; h == "" {
+		return nil, fmt.Errorf("Not a forward rule")
+	}
+
+	return &httputil.ReverseProxy {
+		Director: func(req *http.Request) {
+			req.URL.Scheme = r.Scheme
+			req.URL.Host = r.Forward
+			// req.URL.Path = "/boo1" // TODO JvD - regex to change path here
+		},
+	}, nil
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// ServeHTTP matches the Request with a forward rule and, if found, serves the
+// request with therule's handler. If the rule's secure field is true, it will
+// only allow access if the request has a valid JWT bearer token.
+func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+
+	rule := s.matchRule(req)
+	if rule == nil {
+		http.Error(w, "Not found", http.StatusNotFound)
+		return
+	}
+
+	if rule.Auth {
+		authorized := rule.authorize(w, req)
+		if !authorized {
 			return
 		}
 	}
 
+	if h := rule.handler; h != nil {
+		h.ServeHTTP(w, req)
+		return
+	}
+
 	http.Error(w, "Not Authorized", http.StatusUnauthorized)
 	return
 }
 
-func rejectNoToken(handler http.Handler) http.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusForbidden)
+func (rule *FwdRule) authorize(w http.ResponseWriter, req *http.Request) bool {
```

Comment:
By design, authorization does not check the database to avoid congestion. We use short lived access tokens.